library/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt

```diff
@@ -47,6 +38,86 @@ fun fetch(
 	}
 }
 
+fun performOperation(thingId: String, operationId: String,
+	fillFields: (List<Property>) -> Map<String, Any> = { emptyMap() },
+	onComplete: (Result<Response, Exception>) -> Unit) {
+
+	val thing = graph[thingId]?.value
+
+	thing?.let {
+		val operation = thing.operations[operationId]
+
+		operation?.let {
+			operation.form?.let { form ->
+				if (form.properties.isEmpty()) {
+					requestProperties(form.id) {
+						form.properties = it
+						thing.operations[operationId] = operation
+
+						val attributes = fillFields(it)
+
+						performOperationRequest(operation.target, operation.method, attributes, onComplete)
+					}
+				}
+				else {
+					val attributes = fillFields(form.properties)
+
+					performOperationRequest(operation.target, operation.method, attributes, onComplete)
+				}
+			} ?: performOperationRequest(operation.target, operation.method, emptyMap(), onComplete)
+
+		} ?: onComplete(Result.of(null, { ApioException("Thing $it doesn't have the operation $operationId") }))
+
+	} ?: onComplete(Result.of(null, { ApioException("Thing not found") }))
+}
+
+fun performOperationRequest(url: String, method: String, attributes: Map<String, Any>,
+	onComplete: (Result<Response, Exception>) -> Unit) {
+	launch(UI) {
+		async(CommonPool) {
+
+			val json = Gson().toJson(attributes)
+			val request = createRequest(HttpUrl.parse(url), credential).newBuilder()
```

Comment:
maybe we should extract the request code to a common class/function, I'm not sure.